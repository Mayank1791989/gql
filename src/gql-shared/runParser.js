/* @flow */
import MultilineCharacterStream from './MultilineCharacterStream';
import whileSafe from './whileSafe';
import printTokenState from './printTokenState';
import { type IParser, type TokenState } from './types';

export const BREAK = 'BREAK';

export default function runParser(
  parser: IParser,
  _sourceText: string,
  callback: (
    stream: MultilineCharacterStream,
    state: TokenState,
    style: string,
  ) => null | typeof BREAK,
) {
  const sourceText = _sourceText ? _sourceText : ' ';
  const state = parser.startState();
  const stream = new MultilineCharacterStream(sourceText);

  let style = '';

  let breakByCallback = false;
  whileSafe(
    {
      condition: () => !stream.eof() && !breakByCallback,
      call: () => {
        style = parser.token(stream, state);
        // eslint-disable-next-line callback-return
        const code = callback(stream, state, style);
        if (code === BREAK) {
          breakByCallback = true;
        }
        // if (style === 'js-frag') {
        // console.log(
        //   `position: ${stream.getCurrentPosition()}`,
        //   `start: ${stream.getStartOfToken()}`,
        //   `style: [${style}] token: [${stream.current()}] \n`,
        //   // 'state: ', state, '\n', printTokenState(state),
        // );
        // }
      },
      logOnInfiniteLoop: () => {
        console.log(
          `style: [${style}] token: [${stream.current()}] \n`,
          'state: ',
          state,
          '\n',
          printTokenState(state),
        );
      },
    },
    sourceText.length,
  );
}
