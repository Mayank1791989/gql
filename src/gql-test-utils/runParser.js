/* @flow */
import { type IParser } from 'gql-shared/types';
import MultilineCharacterStream from 'gql-shared/MultilineCharacterStream';

export default function runParser({
  sourceText,
  parser,
}: {
  sourceText: string,
  parser: IParser,
}) {
  const state = parser.startState();
  const stream = new MultilineCharacterStream(sourceText);
  const result = [];
  while (!stream.eof()) {
    const style = parser.token(stream, state);
    result.push({
      start: stream.getStartOfToken(),
      end: stream.getCurrentPosition(),
      text: stream.current(),
      style,
    });
  }

  return result;
}
