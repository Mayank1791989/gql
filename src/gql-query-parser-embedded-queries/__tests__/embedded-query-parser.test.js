/* @flow */
import EmbeddedQueryParserPkg from '../index';
import runParser from 'gql-test-utils/runParser';

test('support single line tag', () => {
  expect(
    runParser({
      sourceText: `
        gql\`
          fragment x on SomeType {
            name
          }
        \`
      `,
      parser: EmbeddedQueryParserPkg({
        start: 'gql`',
        end: '`',
      }).create(),
    }),
  ).toMatchSnapshot();
});

test('support multiline tags', () => {
  expect(
    runParser({
      sourceText: `
        some random text
        graphql request
        """
          fragment x on SomeType {
            name
          }
        """
        some random text
      `,
      parser: EmbeddedQueryParserPkg({
        start: 'graphql request\\s+"""',
        end: '"""',
      }).create(),
    }),
  ).toMatchSnapshot();
});

test('should validate options', () => {
  expect(() => {
    EmbeddedQueryParserPkg('some_wrong_options');
  }).toThrowErrorMatchingInlineSnapshot(`
"INVALID_PARSER_OPTIONS: Error in options passed to 'embedded-query-parser'. 

ParserOptions must be an object

Expected: {|
  start: string;
  end: string;
  allowFragmentWithoutName?: boolean;
  allowFragmentInterpolation?: boolean;
  allowDocumentInterpolation?: boolean;
|}

Actual Value: \\"some_wrong_options\\"

Actual Type: string
"
`);
});
