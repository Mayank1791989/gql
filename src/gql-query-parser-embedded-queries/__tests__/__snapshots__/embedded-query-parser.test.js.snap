// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`support multiline tags 1`] = `
Array [
  Object {
    "end": 61,
    "start": 0,
    "style": "ws-start",
    "text": "
        some random text
        graphql request
        \\"\\"\\"",
  },
  Object {
    "end": 72,
    "start": 61,
    "style": "ws",
    "text": "
          ",
  },
  Object {
    "end": 80,
    "start": 72,
    "style": "keyword",
    "text": "fragment",
  },
  Object {
    "end": 81,
    "start": 80,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 82,
    "start": 81,
    "style": "def",
    "text": "x",
  },
  Object {
    "end": 83,
    "start": 82,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 85,
    "start": 83,
    "style": "keyword",
    "text": "on",
  },
  Object {
    "end": 86,
    "start": 85,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 94,
    "start": 86,
    "style": "atom",
    "text": "SomeType",
  },
  Object {
    "end": 95,
    "start": 94,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 96,
    "start": 95,
    "style": "punctuation",
    "text": "{",
  },
  Object {
    "end": 109,
    "start": 96,
    "style": "ws",
    "text": "
            ",
  },
  Object {
    "end": 113,
    "start": 109,
    "style": "property",
    "text": "name",
  },
  Object {
    "end": 124,
    "start": 113,
    "style": "ws",
    "text": "
          ",
  },
  Object {
    "end": 125,
    "start": 124,
    "style": "punctuation",
    "text": "}",
  },
  Object {
    "end": 134,
    "start": 125,
    "style": "ws",
    "text": "
        ",
  },
  Object {
    "end": 137,
    "start": 134,
    "style": "ws-end",
    "text": "\\"\\"\\"",
  },
  Object {
    "end": 169,
    "start": 137,
    "style": "ws-empty",
    "text": "
        some random text
      ",
  },
]
`;

exports[`support single line tag 1`] = `
Array [
  Object {
    "end": 13,
    "start": 0,
    "style": "ws-start",
    "text": "
        gql\`",
  },
  Object {
    "end": 24,
    "start": 13,
    "style": "ws",
    "text": "
          ",
  },
  Object {
    "end": 32,
    "start": 24,
    "style": "keyword",
    "text": "fragment",
  },
  Object {
    "end": 33,
    "start": 32,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 34,
    "start": 33,
    "style": "def",
    "text": "x",
  },
  Object {
    "end": 35,
    "start": 34,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 37,
    "start": 35,
    "style": "keyword",
    "text": "on",
  },
  Object {
    "end": 38,
    "start": 37,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 46,
    "start": 38,
    "style": "atom",
    "text": "SomeType",
  },
  Object {
    "end": 47,
    "start": 46,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 48,
    "start": 47,
    "style": "punctuation",
    "text": "{",
  },
  Object {
    "end": 61,
    "start": 48,
    "style": "ws",
    "text": "
            ",
  },
  Object {
    "end": 65,
    "start": 61,
    "style": "property",
    "text": "name",
  },
  Object {
    "end": 76,
    "start": 65,
    "style": "ws",
    "text": "
          ",
  },
  Object {
    "end": 77,
    "start": 76,
    "style": "punctuation",
    "text": "}",
  },
  Object {
    "end": 86,
    "start": 77,
    "style": "ws",
    "text": "
        ",
  },
  Object {
    "end": 87,
    "start": 86,
    "style": "ws-end",
    "text": "\`",
  },
  Object {
    "end": 94,
    "start": 87,
    "style": "ws-empty",
    "text": "
      ",
  },
]
`;
